src/core/main.c

```diff
@@ -2263,6 +2265,26 @@ static void fallback_rlimit_memlock(const struct rlimit *saved_rlimit_memlock) {
         arg_default_rlimit[RLIMIT_MEMLOCK] = rl;
 }
 
+static void setenv_manager_environment(void) {
+        char **p;
+
+        /* This destroys arg_manager_environment. */
+
+        STRV_FOREACH(p, arg_manager_environment) {
+                char *e;
+
+                assert_se(e = strchr(*p, '='));
+                *e = '\0';
+
+                log_debug("Setting '%s=%s' in our own environment.", *p, e+1);
+
+                if (setenv(*p, e + 1, true) < 0)
+                        log_warning_errno(errno, "Failed to setenv \"%s\", ignoring: %m", *p);
```

Comment:
urks. patching around in arg_manager_environment? that's evil...

maybe add a helper putenv_by_setenv() which has an API like putenv() but internally uses setenv(), so that our memory stays ours, and it's setenv() that decides where to take the memory from. and that internally uses strndupa() to cut out the variable name, i.e. does stack allocation, because that's easy then, because not in an inner loop

Reply:
I did that, but called it `putenv_dup`.